### PR TITLE
DataForm: enable fields to declare a different layout

### PR DESCRIPTION
**Pull request by louwie17**
## What?

This PR enables a field to declare a layout different from the one defined by the form.

## Why?

In the QuickEdit, we want to display the featured image field using a regular layout, not the panel one. [See related conversation](https://github.com/WordPress/gutenberg/pull/65399).

## How?

- Refactors the existing combined fields API: fields can be a `string`, a `SimpleFormField`, or a `CombinedFormField`.
- Introduces `labelPosition` as a prop to `form` as well as the field (either `SimpleFormField` or `CombinedFormField`). The valid values are `top` (default for `regular` layout), `side` (default for `panel` layout), or `none`.

For example:

```js
const form = {
  fields = [
     'title',
     {
       id: 'featured_media',
       layout: 'regular',
     },
    {
      id: 'status',
      label: __( 'Status & Visibility' ),
      children: [ 'status', 'password' ],
    }
  ]
}
```

How `labelPosition` works with **panel layout** (using the `title` field as an example):

| `top` | `side` (default) | `none` | 
| --- | --- | --- |
| Screenshot 2024-11-18 at 13 14 11 | Screenshot 2024-11-18 at 13 14 46 | Screenshot 2024-11-18 at 13 14 26 |

How `labelPosition` works with **regular layout** (using the `title` field as an example):

| `top` (default) | `side` | `none` |
| --- | --- | --- |
| Screenshot 2024-11-19 at 10 20 42 |  Screenshot 2024-11-18 at 12 36 08 | Screenshot 2024-11-18 at 12 37 51 |

## Testing Instructions

**Test the QuickEdit:**

- With the QuickEdit experiment active, go to the Pages page and open it.
- Verify that the featured image field is rendered as a regular control, while the rest of the fields are rendered as `panel` controls:

Screenshot 2024-11-19 at 14 33 20

**In the storybook, test the type & labelPosition:**

1. Run `npm run storybook:dev` and go to the default DataForm story.
2. The `sticky` field should always be rendered with `labelPosition:side` and `layout: regular`.
3. The rest of the fields should adapt to what the user introduces in the `type` and `labelPosition` values.

### CODE REVIEW

File: packages/dataviews/src/components/dataform-context/index.tsx
Comment (louwie17) [inline packages/dataviews/src/components/dataform-context/index.tsx:1]: Added a DataForm context to store the field definitions. This way I can more easily manage the layout on a field level without having to make sure I drill down the field definitions.

**Comment by oandregal** [comment]
Thanks for keep using the storybook as an example, that's really valuable! 👏 

What do you think of doing the reverse example? In the combinedFields story (`panel` form) configure the reviewer field to use the `inline` layout, for example. Even more ideal would be having the actual use case we want to implement: a feature image.

There's some other challenges with this approach like making the styles work well together (note how the title "Status" has different styles/font size) that make the current example look a bit weird:

Screenshot 2024-10-29 at 10 35 52

File: packages/dataviews/src/components/dataform/stories/index.story.tsx
Comment (oandregal) [inline packages/dataviews/src/components/dataform/stories/index.story.tsx]: I like how expanding the field from string to object allows us to absorb a couple of use cases (grouping fields + controlling field rendering).

Have a couple of thoughts:

- I haven't read the full code yet, but judging from what I tested, I presume by reusing the `id` from an actual field we are able to reuse the field `label`. I understand this may be useful if the only thing a field does is changing its layout. However, how does it look when we want to provide a different title for grouped fields? Do we provide a `label` here as well?

- Another thing I'll be looking in the code review is the logic for stacking fields horizontally or vertically. How does that work without the `direction` config?

File: packages/dataviews/src/types.ts
Comment (oandregal) [inline packages/dataviews/src/types.ts]: Based on what I saw so far, this is the API I was expecting for `FormField`:

```ts
export type FormField =
  | string // The field id
  | {
      id: string; // The field id
      layout?: '...';
      fields?: FormField[]; // Allows grouping fields
  };
```

Some thoughts:

- Why do we need `field` in `FormField`?
- Can the `form.combinedFields` API be removed? Why do we need both FormField and combinedFields?

File: packages/dataviews/src/components/dataform/index.tsx
Comment (oandregal) [inline packages/dataviews/src/components/dataform/index.tsx:20]: I'm mostly interested in getting the API right in the review. However, I've also noticed something related to implementation that we could perhaps simplify.

Can we do the following here (pseudo-code):

```js
// Fields are only the visible fields
// and contain all the information they need including the layout.
const fields = normalizeFields( fields, form );
```

By making sure the normalized fields have all the info they need and it's done as the first step:

- we don't need the context, because the fields are already passed down as they are needed via props, like the `form.fields` work now
- we can remove a lot of code that normalizes in the leaf components. Examples: [formregular](https://github.com/WordPress/gutenberg/blob/a707275562bcb7cd82446759e701cc0de978449e/packages/dataviews/src/dataforms-layouts/regular/index.tsx#L53), [formpanel](https://github.com/WordPress/gutenberg/blob/a707275562bcb7cd82446759e701cc0de978449e/packages/dataviews/src/dataforms-layouts/panel/index.tsx#L175), [dataformlayout](https://github.com/WordPress/gutenberg/blob/a707275562bcb7cd82446759e701cc0de978449e/packages/dataviews/src/dataforms-layouts/data-form-layout.tsx#L36), etc.

File: packages/dataviews/src/types.ts
Comment (oandregal) [reply packages/dataviews/src/types.ts]: I reckon some of this is pre-existing to this PR. If you think it's too much for this PR, we could do a PR to simplify the implementation first, and then rebase this PR to use that.

File: packages/dataviews/src/dataforms-layouts/index.tsx
Comment (oandregal) [inline packages/dataviews/src/dataforms-layouts/index.tsx:7]: Don't we have a `FORM_LAYOUT` constant above? Why do we need both?

File: packages/dataviews/src/types.ts
Comment (louwie17) [reply packages/dataviews/src/types.ts]: > Why do we need field in FormField?

Actually my thought was to use `field` for the `field` id instead of `id`, as `field` is more clear. We did probably want to replace `fields` with `children` to not make it to confusing.
Mostly for this scenario, where what would you do if you wanted to display a horizontal column layout within a panel. Or any horizontal layout in general. In this case we are just grouping fields, without needing to reference a specific field.
It may look like this:
```
{
  id: "status",
  layout: "panel",
  fields: [
      {
          layout: 'horizontal' // if we go this route
          fields: [ 'status', 'password' ]
      }
  ]
}
```
We may still want to add an `id` here as a `key`, although we could also generate one.

> Can the form.combinedFields API be removed? Why do we need both FormField and combinedFields?

Yeah `combinedFields` can be removed, as this API replaces that functionality. I was thinking of doing that as a follow up. But happy to remove it as part of this PR.

File: packages/dataviews/src/components/dataform/stories/index.story.tsx
Comment (louwie17) [reply packages/dataviews/src/components/dataform/stories/index.story.tsx]: > I haven't read the full code yet, but judging from what I tested, I presume by reusing the id from an actual field we are able to reuse the field label. I understand this may be useful if the only thing a field does is changing its layout. However, how does it look when we want to provide a different title for grouped fields? Do we provide a label here as well?

Correct, we use the `label` of the referenced field. I agree this is not ideal for the group scenario that you mentioned and we may want to add additional items to the `FormField` config.
We could add a `label` option and default to the field `label` if a label is not mentioned.

> Another thing I'll be looking in the code review is the logic for stacking fields horizontally or vertically. How does that work without the direction config?

Instead of a `direction` config, I was thinking of adding a `horizontal` and `group` layout ( where vertical is the default ). Similar to https://jsonforms.io/examples/layouts/ 

For both of these questions the answer/implementation is a bit easier for the regular and inline layouts, but the `panel` layout is a bit of an exception, given it contains both the `render` and the `edit` view.

For example, is there a use case for grouping fields within a panel, where the panel `render` view shows both field values, but still renders the fields in a single dropdown.
This may very well be overthinking it though, and I am not sure if we have this scenario currently present within WP.

File: packages/dataviews/src/dataforms-layouts/index.tsx
Comment (louwie17) [reply packages/dataviews/src/dataforms-layouts/index.tsx:7]: That is still referring to the current form layout that determines it for the entire form.
This can be removed if we go ahead with the form field layout instead.

File: packages/dataviews/src/components/dataform/index.tsx
Comment (louwie17) [reply packages/dataviews/src/components/dataform/index.tsx:20]: If I understand you correctly, you mean replacing the `form.fields` array with the normalized fields already. So `[ "status" ]` would turn into `[ { ...status field definition } ]`?

I am ok with this approach, but initially avoided it because of the nesting possibility. This means we will have to traverse through the entire field tree initially already, instead of having the react components deal with this naturally.

> we can remove a lot of code that normalizes in the leaf components. Examples: [formregular](https://github.com/WordPress/gutenberg/blob/a707275562bcb7cd82446759e701cc0de978449e/packages/dataviews/src/dataforms-layouts/regular/index.tsx#L53), [formpanel](https://github.com/WordPress/gutenberg/blob/a707275562bcb7cd82446759e701cc0de978449e/packages/dataviews/src/dataforms-layouts/panel/index.tsx#L175), [dataformlayout](https://github.com/WordPress/gutenberg/blob/a707275562bcb7cd82446759e701cc0de978449e/packages/dataviews/src/dataforms-layouts/data-form-layout.tsx#L36), etc.

With this new implementation `FormRegular` and `FormPanel` can actually be removed as they are replaced by `FormRegularField` and `FormPanelField`.
`DataFormLayout` would be the only exception.

File: packages/dataviews/src/components/dataform/stories/index.story.tsx
Comment (louwie17) [inline packages/dataviews/src/components/dataform/stories/index.story.tsx:123]: Added sticky example, which flows nicer for the inline example.

File: packages/dataviews/src/dataforms-layouts/panel/index.tsx
Comment (gigitux) [inline packages/dataviews/src/dataforms-layouts/panel/index.tsx]: Could it be useful to keep `_x` to give more context to translators?

File: packages/dataviews/src/components/dataform-context/index.tsx
Comment (louwie17) [inline packages/dataviews/src/components/dataform-context/index.tsx]: I added the `label` option to the `FormField` definition, which overwrites the field definition label if provided.
I did this for two reasons:
- This way the individual layouts won't need to worry about this
- The regular layout doesn't handle the label rendering and renders the Edit component directly, which would require us to overwrite the label there anyway.

File: packages/dataviews/src/dataforms-layouts/data-form-layout.tsx
Comment (louwie17) [inline packages/dataviews/src/dataforms-layouts/data-form-layout.tsx]: Added the `isVisible` logic here

File: packages/dataviews/src/dataforms-layouts/get-visible-fields.ts
Comment (louwie17) [inline packages/dataviews/src/dataforms-layouts/get-visible-fields.ts:23]: Removed the combined fields related stuff

**Comment by louwie17** [comment]
Thanks for the initial review @oandregal, I did some clean up and addressed some of your suggestions. I also left some replies to your above comments.

> Thanks for keep using the storybook as an example, that's really valuable! 👏
> 
> What do you think of doing the reverse example? In the combinedFields story (`panel` form) configure the reviewer field to use the `inline` layout, for example. Even more ideal would be having the actual use case we want to implement: a feature image.
> 
> There's some other challenges with this approach like making the styles work well together (note how the title "Status" has different styles/font size) that make the current example look a bit weird:

Good suggestion on the reverse, I end up creating a simple `Sticky` field example within the storybook for the default example. I also updated the logic a bit, so that it only uses the `inline` layout when in the `panel` or `inline` view. If you switch to `regular` it will switch it to `regular` as well.
Screenshot 2024-10-31 at 2 47 13 PM

File: packages/dataviews/src/components/dataform-context/index.tsx
Comment (gigitux) [inline packages/dataviews/src/components/dataform-context/index.tsx]: ~~Can't we move this logic to `normalizeFields` function? In this way, we can get rid of this context~~

I misunderstood this code!

**Review by gigitux — COMMENTED** [review]
I left two comments: one is straightforward. For the second one, I'm not sure why the context is needed 🤔

File: packages/dataviews/src/dataforms-layouts/panel/index.tsx
Comment (louwie17) [reply packages/dataviews/src/dataforms-layouts/panel/index.tsx]: Aaah yes, this may of been my mistake when fixing some merge conflicts, nice catch!

File: packages/edit-site/src/components/post-edit/index.js
Comment (oandregal) [inline packages/edit-site/src/components/post-edit/index.js]:
```suggestion
```

**Comment by oandregal** [comment]
Thanks for updating!

It looks like the initial status for the layout of the `sticky` field is not properly set:

https://github.com/user-attachments/assets/3d52d70a-d565-4049-8fb1-01be4146da92

File: packages/dataviews/src/types.ts
Comment (oandregal) [reply packages/dataviews/src/types.ts]: > We did probably want to replace fields with children to not make it to confusing.

Agree to this, it'd make the code clearer.

Why the `regular` layout can't take children but the `panel` layout can? The existing combinedFields API worked for both.

Comment (oandregal) [reply packages/dataviews/src/types.ts]: Another thought is: why only the fields that are defined via object in `form.fields` should have the `layout` property? It looks like this is optimizing for the current use case (combining status + password), but I wonder if the layout should be part of the Fields API instead?

File: packages/dataviews/src/components/dataform/index.tsx
Comment (oandregal) [reply packages/dataviews/src/components/dataform/index.tsx:20]: As it is, normalization is spread across multiple layers/components: normalizeField utility, the dataform context (for the label), the dataform layout (for the field layout).

What do you think of 1) making `layout` part of the Fields API, 2) adding support for `children` in both regular&panel, 3) centralizing everything in `normalizeFields` (removing the context)?

**Comment by oandregal** [comment]
This is nice, thanks for the effort!

The main blocker I see is adding support for `children` in the `regular` panel like we had with the combinedFields API.

File: packages/dataviews/src/types.ts
Comment (youknowriad) [inline packages/dataviews/src/types.ts]: What's the difference between RegularField and InlineField?

Comment (louwie17) [reply packages/dataviews/src/types.ts]: Regular field:
Screenshot 2024-11-04 at 1 27 05 PM

Inline field:
Screenshot 2024-11-04 at 1 27 18 PM

Inline has a column layout similar to panel field, but displays the actual edit field on the right side.
Happy to change the name if there is a better alternative?

Comment (youknowriad) [reply packages/dataviews/src/types.ts]: Here's my take on this, I think we only have two kind of layouts: `inline` (or regular) and `panel` and both can support labelPosition as "side" or "top". It's kind of similar to the "labelPosition" prop in React Spectrum https://react-spectrum.adobe.com/react-spectrum/TextField.html#label-alignment-and-position

Comment (louwie17) [reply packages/dataviews/src/types.ts]: 👍 I like that idea as well, lets go with that.
I will just have to see how the `top` label position will look like for the panel view, given just moving the button below the label may be a bit strange:
Screenshot 2024-11-04 at 2 50 42 PM

I guess wrapping it with the `BaseControl` could suffice:
Screenshot 2024-11-04 at 2 52 08 PM

Comment (youknowriad) [reply packages/dataviews/src/types.ts]: If the "button" is labeled, Adding `BaseControl` might create two labels (so not entirely correct a11y wise), so maybe the visuals should match what you shared above but it should just be adhoc matching.

File: packages/dataviews/src/components/dataform-context/index.tsx
Comment (youknowriad) [inline packages/dataviews/src/components/dataform-context/index.tsx]: In the DataViewsContext, we store the `fields` directly in the context, while here we provide an abstraction function. I wonder if we should be consistent and just keep `fields` here.

File: packages/dataviews/src/dataforms-layouts/panel/index.tsx
Comment (louwie17) [inline packages/dataviews/src/dataforms-layouts/panel/index.tsx:62]: Moved the `Dropdown` code to its own component to better handle the `labelPosition` in the `FormPanelField` component.

File: packages/dataviews/src/components/dataform-context/index.tsx
Comment (youknowriad) [inline packages/dataviews/src/components/dataform-context/index.tsx]: This is weird to me, why are we overriding the definition. What "type" are we returning here.

File: packages/dataviews/src/components/dataform/index.tsx
Comment (youknowriad) [reply packages/dataviews/src/components/dataform/index.tsx:20]: @oandregal it's important to distinguish `form.fields` from the `fields` prop which represents the "schema" of our data basically. 

The function being called here, normalized the "schema" and doesn't touch `form.fields` at all. We could obviously introduce `normalizeForm` function or something (which is I believe what you're suggesting) but it's important to keep these things distinct. `normalizeFields` is a function that is common between DataViews and DataForms.

File: packages/dataviews/src/types.ts
Comment (youknowriad) [inline packages/dataviews/src/types.ts]: It's weird to me that "children" is part of FormField. In other words, this would make more sense to me:

```js 
type SimpleFormField = {
   id: string; // could be named "field" as well if we want to match "filters"
   layout: 'regular' | 'panel';
   labelPosition: 'side' | 'top';
}

type CombinedFormField = {
   id: 'string'; // actually not sure this is necessary as it's just a random string but maybe it can be useful 
   label: 'string';
   layout: 'regular' | 'panel';
   children: Array<FormField>
}

type FormField = string | SimpleFormField | CombinedFormField;
```

It is possible that I'm misunderstanding combined fields, but for me, it's weird to make a field children of another field

Comment (youknowriad) [reply packages/dataviews/src/types.ts]: I see the example of "status" with "password".  Is "password" a child of "status" or we're creating a "virtual combined field" that combines both in a single panel. I feel the latter is closer to reality. 

I guess the main difference is that I'm forcing the "label" and "children" to be required for "combined fields" and I'm removing "label" for regular fields because the label is inherited from the field directly and also preventing "regular" fields to have childre.

Comment (youknowriad) [reply packages/dataviews/src/types.ts]: I might be getting into details that don't matter much at the moment, so feel free to take this comment with a grain of salt (fine to ignore for now)

File: packages/dataviews/src/components/dataform-context/index.tsx
Comment (louwie17) [reply packages/dataviews/src/components/dataform-context/index.tsx]: Yeah I am happy to use `fields` here instead for consistency.

File: packages/dataviews/src/components/dataform-context/index.tsx
Comment (louwie17) [reply packages/dataviews/src/components/dataform-context/index.tsx]: We are returning the `NormalizedField` type here. I explained the reasoning for going this route as part of this comment: https://github.com/WordPress/gutenberg/pull/66531#discussion_r1824481795
But with [this change](https://github.com/WordPress/gutenberg/pull/66531#discussion_r1829334630), some of the `NormalizedField` retrieval logic will move back to the layout, so I can keep it there. Which would be better to avoid confusion.

File: packages/dataviews/src/types.ts
Comment (louwie17) [reply packages/dataviews/src/types.ts]: No worries of digging into the details, this part of the API is what probably needs the most fine tuning anyhow.

> It is possible that I'm misunderstanding combined fields, but for me, it's weird to make a field children of another field

True, not ideal, and I like your suggestion of separating it out again. I was running into this in the code as well, in particular with the `panel` layout as I was adding the field it-self to its children: https://github.com/WordPress/gutenberg/blob/ddf996a583b00052e1e8e0e031c616eb026c0f37/packages/dataviews/src/dataforms-layouts/panel/index.tsx#L83

The panel layout is actually what I have been struggling with the most in regards to this, as it goes a bit against the grain compared to the regular layout.
For example in the regular form, it would make sense to wrap children up in combined fields, as all is required is a label and a wrapper in essence.
The panel requires the label, but also the render method, but in the status case we display a single `value` and render multiple edit fields ( already requiring some custom logic ).

So with using the above `CombinedFormField` and `panel` layout, we may still need a reference to a `render` method. In which case I defaulted to the field referenced as part of the `id`. But maybe the first one in the `children` list would suffice?

_Sorry that may came across confusing._

> I see the example of "status" with "password". Is "password" a child of "status" or we're creating a "virtual combined field" that combines both in a single panel. I feel the latter is closer to reality.

Its the latter, a "virtual combined field". You can see my slight frustration above with how this exactly fits within the current API structure.

Comment (youknowriad) [reply packages/dataviews/src/types.ts]: > So with using the above CombinedFormField and panel layout, we may still need a reference to a render method. In which case I defaulted to the field referenced as part of the id. But maybe the first one in the children list would suffice?
> Sorry that may came across confusing.

No it actually make sense. I'm fine with using the first field to start with (I like starting small), but it does highlight that maybe at some point a custom render function receiving the children could be defined (like concatenating the values or things like that) :)

File: packages/dataviews/src/components/dataform-context/index.tsx
Comment (louwie17) [reply packages/dataviews/src/components/dataform-context/index.tsx]: I removed this.

File: packages/dataviews/src/types.ts
Comment (louwie17) [reply packages/dataviews/src/types.ts]: I added the above type as part of this commit: https://github.com/WordPress/gutenberg/pull/66531/commits/21f7f9ed85d40f1d45436ad32dce75887727a47e
It works ok, at the moment it did require a few type checks in additional places. So there is room for cleanup, but that could be also be done as a follow up IMO.

Currently I have both layout fields ( regular & panel ) handle the combined fields. But I wonder if I can create a better abstraction for combined fields that would use the regular & panel under the hood. Again easier for the regular view instead of the panel view.

Also the only real difference for the panel view is that we use the first child for the `render` view. This is not a 100% clear from the code, and would require some clarification.

**Comment by louwie17** [comment]
Alright this should be ready for a re-review, I think it does require some additional clean up, but it does work with the new layout of the API if we are on board with that: https://github.com/WordPress/gutenberg/pull/66531/files#diff-05cd2939bbf237a5dbcd116a0da9c9019b05ec1ceaa9d429c5fc872f149f63c4R528-R542
Maybe the big question is, do we want to allow combined fields to be nested of each other?
Ex:
```
{
   // combined field
   children: [
     field1,
     field2,
     {
       // combined field
      children: [ field3, field4 ]
     }
   ]
}
```

If the API structure does seem good, than I would be more inclined to merge this if it in general looks good and address any clean up as follow up ( but happy to change my mind on this ).

File: packages/dataviews/src/components/dataform/index.tsx
Comment (oandregal) [reply packages/dataviews/src/components/dataform/index.tsx:20]: I agree `normalizeFields` is common across DataViews/DataForm so it shouldn't have the `form` argument and we may need another function or just have some code here. That's fine. My point was about simplifying the implementation. Let me try to rephrase a bit to see if it comes across:

The way I see it, `form.fields` is a filter on `fields` (it's also creating new ones, for example, combined fields). So, if we normalize the fields as early as possible, all internal components can just work with the Fields API (`fields`).

As I mentioned elsewhere, this conversation is not a blocker to for me. I'm mostly interested in getting the API right.

File: packages/dataviews/src/components/dataform/stories/index.story.tsx
Comment (oandregal) [inline packages/dataviews/src/components/dataform/stories/index.story.tsx]: I'm a bit confused as to why we need this? Isn't this part of what DataForm does internally? The other consumer (edit-site post-edit) doesn't require it (and it shouldn't).

Comment (louwie17) [reply packages/dataviews/src/components/dataform/stories/index.story.tsx]: This is not necessary by consumers, only if they need to use a `labelPosition` that isn't the default.
I only added this in Storybook because it makes it easier to switch between the layout & labelPosition argument types, mostly for display purposes.
If this is confusing, what I could do is only use a single field as an example for the `labelPosition` versus the entire form.

File: packages/dataviews/src/components/dataform/index.tsx
Comment (louwie17) [reply packages/dataviews/src/components/dataform/index.tsx:20]: > The way I see it, form.fields is a filter on fields (it's also creating new ones, for example, combined fields). So, if we normalize the fields as early as possible, all internal components can just work with the Fields API (fields).

That makes sense, my only thought would be if there could be a case where "a field definition is needed for a field that is not in the rendered list ( `form.fields` )?"
Like:
- A conditional field
     - Should not be needed because of the `isVisible` function, and updating `form.fields` on the fly wouldn't be necessary anymore.
- A field that requires data of another field definition and may need to use `getValue` internally ( the slug comes to mind, as it would be auto generated from the title ). 
     - This would be avoided if we still expose the `fields` definitions anyway.

**Comment by oandregal** [comment]
This PR also fixes an issue I ran into while working with the status field at https://github.com/WordPress/gutenberg/pull/66937#issuecomment-2470804044

What's the status of this PR? Is it ready for a final review/approval?

**Comment by louwie17** [comment]
> This PR also fixes an issue I ran into while working with the status field at [#66937 (comment)](https://github.com/WordPress/gutenberg/pull/66937#issuecomment-2470804044)

👍 

> What's the status of this PR? Is it ready for a final review/approval?

Yeah it should be ready for a final review/approval.
I did [leave a reply to one of your last comments](https://github.com/WordPress/gutenberg/pull/66531#discussion_r1832628579) in a change within the storybook file, to see if you wanted me to change/clarify that before this gets merged.

**Comment by oandregal** [comment]
I've taken a look and pushed some minimal changes. I think this work is best highlighted by actually implementing what we are trying to do: render the image edit control inline in the panel layout. So I've configured it to do so:

Screenshot 2024-11-15 at 14 44 38

There's some details to figure out about `labelPosition`. In particular, I think we need `labelPosition: none`. Note that the example is working as expected (no label for `featured_media`) but for the wrong reasons: the field doesn't actually render any label. I think we have to 1) render a label for it and 2) let the form configure it (in this case, we want "none").

For now, I've just focused on polishing the API: it's working as expected other than what I mentioned. After that, I'd like to do a second review focused on the implementation details.

**Comment by oandregal** [comment]
Pushed https://github.com/WordPress/gutenberg/pull/66531/commits/f64d95f78c55c617cd5a9a16a9cc6aaee5facd0a and https://github.com/WordPress/gutenberg/pull/66531/commits/50d81940f69fe325413d7aaba331709c3f6775bd to implement `labelPosition: none` for the panel & regular layouts.

There are some use cases where we may need to position the label in a different place than the default ([sticky](https://github.com/WordPress/gutenberg/pull/62782), [settings](https://github.com/WordPress/gutenberg/issues/63624)). This may need more work but, for now, this PR implements `top`, `side` and `none` values.

How `labelPosition` works with **panel layout** (using the `title` field as an example):

| `top` | `side` (default) | `none` | 
| --- | --- | --- |
| Screenshot 2024-11-18 at 13 14 11 | Screenshot 2024-11-18 at 13 14 46 | Screenshot 2024-11-18 at 13 14 26 |

How `labelPosition` works with **regular layout** (using the `title` field as an example):

| `top` (default) | `side` | `none` |
| --- | --- | --- |
| Screenshot 2024-11-19 at 10 20 42 |  Screenshot 2024-11-18 at 12 36 08 | Screenshot 2024-11-18 at 12 37 51 |

**Comment by oandregal** [comment]
Finalized implementing the `labelPosition`: it now supports `none`. Additionally, the form can also define a default `labelPosition`, which will be used if the field doesn't define one.

Also updated the description to better reflect what this PR does.

File: packages/dataviews/src/components/dataform/stories/index.story.tsx
Comment (oandregal) [reply packages/dataviews/src/components/dataform/stories/index.story.tsx]: With the latest changes (adding `labelPosition` to `form`) the `toFormField` utility is not necessary.

**Review by oandregal — APPROVED** [review]

**Comment by youknowriad** [comment]
I'm really excited about this one. I think this is where the power of the DataForm starts to show up. Thank you all for the collaboration.

File: packages/dataviews/src/dataforms-layouts/panel/index.tsx
Comment (louwie17) [inline packages/dataviews/src/dataforms-layouts/panel/index.tsx:89]: Currently we have this same logic in the `panel` and `regular` layouts, which is very similar to the `normalizeFormFields` function ( I believe I implemented this ). I think we could call that instead:
```
return normalizeFormFields( {
  type: 'regular' as const,
  fields: field.children
} );
```

This is something I can do as a follow up PR as this works fine as is.

**Pull request merged**
